File: python-client/docker_consumer.py
"""Docker consumer — simulated per-item process; env from entrypoint-consumer.sh."""

from __future__ import annotations

import asyncio
import os
import random

from cupidmq import CupidMQ


def _env(name: str, default: str = "") -> str:
    return os.environ.get(name, default)


async def main() -> None:
    master = _env("CUPIDMQ_MASTER", "127.0.0.1:9750")
    data_addr = _env("CUPIDMQ_DATA_ADDR")
    if not data_addr:
        raise SystemExit("CUPIDMQ_DATA_ADDR required")

    bind_addr = _env("CUPIDMQ_BIND_ADDR") or None
    tag = _env("CUPIDMQ_CONSUMER_TAG")
    process_min_ms = float(_env("CUPIDMQ_PROCESS_MS_MIN", "2"))
    process_max_ms = max(process_min_ms, float(_env("CUPIDMQ_PROCESS_MS_MAX", "16")))

    client = CupidMQ.consumer(
        master,
        data_addr=data_addr,
        bind_addr=bind_addr,
        consumer_tag=tag,
    )
    cc = client._config.consumer  # noqa: SLF001 — startup log

    print(
        f"docker-consumer-py: master={master} advertise={cc.data_addr} "
        f"bind={cc.effective_bind_addr} tag={tag} "
        f"process={process_min_ms:.0f}-{process_max_ms:.0f}ms/item",
        flush=True,
    )

    async with client:
        async for batch in client.consume():
            for item in batch:
                delay_s = random.uniform(process_min_ms, process_max_ms) / 1000.0
                if delay_s > 0:
                    await asyncio.sleep(delay_s)
                _ = len(item)


if __name__ == "__main__":
    asyncio.run(main())

File: master/examples/docker-consumer.rs
//! Docker / Compose consumer — config via env (see docker/entrypoint-consumer.sh).

use anyhow::{Context, Result};
use cupidmq::consumer::ConsumerConfig;
use cupidmq::CupidMQ;
use rand::Rng;
use std::env;
use std::time::Duration;

fn required(name: &str) -> Result<String> {
    env::var(name).with_context(|| format!("{name} required"))
}

#[tokio::main]
async fn main() -> Result<()> {
    let master = env::var("CUPIDMQ_MASTER").unwrap_or_else(|_| "127.0.0.1:9750".into());
    let data_addr = required("CUPIDMQ_DATA_ADDR")?;

    let mut config = ConsumerConfig::new(master, data_addr)?;
    if let Ok(bind) = env::var("CUPIDMQ_BIND_ADDR") {
        config = config.bind_addr(bind)?;
    }
    if let Ok(tag) = env::var("CUPIDMQ_CONSUMER_TAG") {
        config = config.consumer_tag(tag);
    }

    let bind = config.resolve_bind()?;
    eprintln!(
        "docker-consumer: advertise={} bind={}:{}",
        config.data_addr, bind.0, bind.1
    );

    let process_min_ms: u64 = env::var("CUPIDMQ_PROCESS_MS_MIN")
        .ok()
        .and_then(|v| v.parse().ok())
        .unwrap_or(2);
    let process_max_ms: u64 = env::var("CUPIDMQ_PROCESS_MS_MAX")
        .ok()
        .and_then(|v| v.parse().ok())
        .unwrap_or(16)
        .max(process_min_ms);

    eprintln!(
        "docker-consumer: process {process_min_ms}-{process_max_ms}ms/item",
    );

    let mut client = CupidMQ::connect_consumer(config).await?;

    client
        .consume(move |batch| async move {
            let mut rng = rand::thread_rng();
            for item in &batch {
                let ms = rng.gen_range(process_min_ms..=process_max_ms);
                if ms > 0 {
                    tokio::time::sleep(Duration::from_millis(ms)).await;
                }
                let _ = item.len();
            }
            Ok(())
        })
        .await?;

    Ok(())
}
